### PR TITLE
fix cluster/log-dump/log-dump.sh shellcheck failures

### PR DESCRIPTION
**Pull request by BenTheElder**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**: Fixes shellcheck failures (again) in cluster/log-dump.sh

**Which issue(s) this PR fixes**:

ref: #72956 

**Special notes for your reviewer**: This does *not* include changes that previously caused a revert, we've since merged https://github.com/kubernetes/kubernetes/pull/85913 with different changes to that part.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @jkaniuk
hoping we can get this done right this time, we're down to 9 scripts left.

File: cluster/log-dump/log-dump.sh
Comment (joakimr-axis) [inline cluster/log-dump/log-dump.sh]: Quotes are superfluous when assigning from other variable.
You can test that e.g. by typing:

$ a='foo bar'
$ b=$a
$ echo $b
foo bar
$

**Comment by jkaniuk** [comment]
> cc @jkaniuk
> hoping we can get this done right this time, we're down to 9 scripts left.

@BenTheElder - sorry for a long response, lost track of it somehow. LGTM from scalability perspective.
/lgtm

**Comment by jkaniuk** [comment]
but
/hold
as there was another comment to address

**Comment by BenTheElder** [comment]
I'd rather leave it for now for consistency / the code is valid enough & bigger fish to fry right now.
We can change that in a follow up
/hold cancel

**Comment by jkaniuk** [comment]
Sure.
PR still needs rebasing.

**Comment by BenTheElder** [comment]
/lgtm cancel
failing shellcheck again

File: cluster/log-dump/log-dump.sh
Comment (joakimr-axis) [inline cluster/log-dump/log-dump.sh:109]: Wouldn't you want something like

`while IFS= read -r line; do files+=("$line"); done <<< "$3"`

here to populate the `files` array?

**Comment by jkaniuk** [comment]
/reopen

@BenTheElder - do you plan to continue with it?
we've invested quite a lot in it already (including 5k scalability issue debugging)

File: cluster/log-dump/log-dump.sh
Comment (joakimr-axis) [inline cluster/log-dump/log-dump.sh:146]: This is where shellcheck currently throws you an error; `files` is assigned as a string but used as an array.

**Comment by BenTheElder** [comment]
I've pushed an admittedly low effort fixup. I can take a closer look later.

File: cluster/log-dump/log-dump.sh
Comment (spiffxp) [inline cluster/log-dump/log-dump.sh:492]: Unless I'm missing where this gets initialized? Follows pattern used by `NON_LOGEXPORTED_NODES` above
```suggestion
    NODE_NAMES=()
    while IFS='' read -r line; do NODE_NAMES+=("$line"); done < <(log_dump_custom_get_instances node)
```

**Review by spiffxp — COMMENTED** [review]
One nit otherwise LGTM

File: cluster/log-dump/log-dump.sh
Comment (joakimr-axis) [reply cluster/log-dump/log-dump.sh:492]: Very good suggestion, thanks a lot! Will fix immediately.

Comment (BenTheElder) [reply cluster/log-dump/log-dump.sh:493]: fixed

**Comment by BenTheElder** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/88349/pull-kubernetes-conformance-kind-ga-only-parallel/1299249213367914496
> wrapper.sh] [TEST] Running Test Command: `bash -c curl -sSL https://kind.sigs.k8s.io/dl/latest/linux-amd64.tgz | tar xvfz - -C "${PATH%%:*}/" && e2e-k8s.sh` ...
> curl: (60) SSL: no alternative certificate subject name matches target host name 'kind.sigs.k8s.io'
> More details here: https://curl.haxx.se/docs/sslcerts.html

wtf? https://kind.sigs.k8s.io/ has a valid letsencrypt cert ...
/retest

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm
/milestone v1.20

**Comment by mm4tt** [comment]
This PR broke log dumping. 

There are no master component (etcd, kube-apiserver, kube-scheduler, etc.) logs dumped in the presubmits of this PR - https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/88349/pull-kubernetes-e2e-gce-100-performance/1301152255931584512/
Master logs - https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/88349/pull-kubernetes-e2e-gce-100-performance/1301152255931584512/artifacts/e2e-88349-95a39-master/
Dumping the logs fails with
```
I0902 14:23:02.917] Copying 'kube-apiserver.log kube-apiserver-audit.log kube-scheduler.log kube-controller-manager.log etcd.log etcd-events.log glbc.log cluster-autoscaler.log kube-addon-manager.log konnectivity-server.log fluentd.log kubelet.cov cl2-* startupscript.log' from e2e-88349-95a39-master
W0902 14:23:04.068] 
W0902 14:23:04.069] Specify --start=45420 in the next get-serial-port-output invocation to get only the new output starting from here.
W0902 14:23:05.339] load pubkey "/workspace/.ssh/google_compute_engine": invalid format
W0902 14:23:05.858] scp: {/var/log/kube-apiserver.log: No such file or directory
W0902 14:23:05.859] scp: kube-apiserver-audit.log: No such file or directory
W0902 14:23:05.859] scp: kube-scheduler.log: No such file or directory
W0902 14:23:05.859] scp: kube-controller-manager.log: No such file or directory
W0902 14:23:05.859] scp: etcd.log: No such file or directory
W0902 14:23:05.859] scp: etcd-events.log: No such file or directory
W0902 14:23:05.859] scp: glbc.log: No such file or directory
W0902 14:23:05.860] scp: cluster-autoscaler.log: No such file or directory
W0902 14:23:05.860] scp: kube-addon-manager.log: No such file or directory
W0902 14:23:05.860] scp: konnectivity-server.log: No such file or directory
W0902 14:23:05.860] scp: fluentd.log: No such file or directory
W0902 14:23:05.860] scp: kubelet.cov: No such file or directory
W0902 14:23:05.861] scp: cl2-*: No such file or directory
W0902 14:23:05.861] scp: startupscript.log*,}: No such file or directory
```


It worked before this PR was merged, e.g. https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1300905180866285571 and https://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1300905180866285571/artifacts/e2e-big-master/


Will send a revert shortly

**Comment by joakimr-axis** [comment]
I think I found the error, @BenTheElder:
In the `copy-logs-from-node` function

`while IFS= read -r line; do files+=("$line"); done <<< "$3"`

will result in an array with 1 entry only (which is a string with all the _N_ filenames) instead of an array with _N_ entries which is what the old code produces. One option would/could be

```
local -r node="${1}"
local -r dir="${2}"
# We want to split $3 on whitespace
# shellcheck disable=SC2086
read -r -a files <<< $3
```

(Or set a shellcheck disable for the old code and call it a day.)

**Comment by joakimr-axis** [comment]
I would personally prefer to, in the `save-logs` function, let `files` (your `save_files`) be an array instead, and thus call `copy-logs-from-node` with

`copy-logs-from-node "${node_name}" "${dir}" "${save_files[@]}"`

and update the assignment of `files` in `copy-logs-from-node` like this:

```
local -r node="${1}"
local -r dir="${2}"
shift
shift
local -r files=("$@")
```

That way filenames with eventual whitespaces would be properly handled too.

**Comment by joakimr-axis** [comment]
Simple fix to mitigate the issue: #94762 (that will handle whitespaces in the same manner as the current code).

**Pull request merged**
